lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/TaskManager.cs

```diff
@@ -0,0 +1,500 @@
+        {
+            var errorMessage = failedTask.Message;
+            switch (errorMessage)
+            {
+                case TaskAppError:
+                    _numberOfAppError++;
+                    return TaskStateEvent.FailedTaskAppError;
+                case TaskSystemError:
+                    return TaskStateEvent.FailedTaskSystemError;
+                case TaskGroupCommunicationError:
+                    return TaskStateEvent.FailedTaskCommunicationError;
+                default:
+                    return TaskStateEvent.FailedTaskSystemError;
+            }
+        }
+
+        /// <summary>
+        /// Returns the number of application error caused by FailedTask
+        /// </summary>
+        /// <returns></returns>
+        public int NumberOfAppError()
+        {
+            return _numberOfAppError;
+        }
+
+        /// <summary>
+        /// Checks if all the tasks are in final states
+        /// </summary>
+        /// <returns></returns>
+        internal bool AllInFinalState()
+        {
+            return _tasks.All(t => t.Value.Item1.IsFinalState());
+        }
+
+        /// <summary>
+        /// Gets current state of the task
+        /// </summary>
+        /// <param name="taskId"></param>
+        /// <returns></returns>
+        internal TaskState TaskState(string taskId)
+        {
+            if (!_tasks.ContainsKey(taskId))
+            {
+                var msg = string.Format(CultureInfo.InvariantCulture, "The task [{0}] doesn't exist.", taskId);
+                Exceptions.Throw(new IMRUSystemException(msg), Logger);
+            }
+
+            var taskInfo = GetTaskInfo(taskId);
+            return taskInfo.Item1.CurrentState;
+        }
+
+        /// <summary>
+        /// Checks if all the tasks are in the state specified. 
+        /// For example, passing TaskState.TaskRunning to check if all the tasks are in TaskRunning state
+        /// </summary>
+        /// <param name="taskState"></param>
+        /// <returns></returns>
+        internal bool AllInTheState(TaskState taskState)
+        {
+            return _tasks.All(t => t.Value.Item1.CurrentState == taskState);
+        }
+
+        /// <summary>
+        /// Submit all the tasks
+        /// Tasks will be submitted after all the tasks are added in the collection and master task exists
+        /// IMRUSystemException will be thrown if not all the tasks are added or if there is no master task
+        /// </summary>
+        internal void SubmitTasks()
+        {
+            if (NumberOfTasks < _totalExpectedTasks || !MasterTaskExists())
+            {
+                string msg = string.Format("Trying to submit tasks but either master task doesn't exist or number of tasks [{0}] is smaller than expected number of tasks [{1}].", NumberOfTasks, _totalExpectedTasks);
+                Exceptions.Throw(new IMRUSystemException(msg), Logger);
+            }
+
+            StartTask(_masterTaskId);
```

Comment:
I don't think it is necessary as all the nodes are independent. We used to start master first then others I guess it was because broadcast starts from master. But since all the nodes have to wait until all others finish the initialization, it actually doesn't matter.